# Koneoppiminen: Tehtävä 3: Naiivi Bayes-luokittelija (max. 9p)

Aiheena **Naiivi Bayes -luokittelija**: Naiivit Bayes on ohjattu oppimismenetelmä, joka perustuu Bayesin lauseen soveltamiseen "naiivilla" oletuksella ehdollisesta riippumattomuudesta jokaisen ominaisuuden välillä. Huolimatta näennäisesti yksinkertaistetuista oletuksistaan, naiivit Bayes-luokittelijat ovat toimineet hyvin monissa todellisissa tilanteissa, kuuluisa dokumenttien luokittelussa ja roskapostin suodatuksessa.

**Muista** kirjoittaa alle omat tietosi.

In [1]:
# Kirjoita tähän tietosi!
student_name = 'Jere Soininen'
student_id = 'AB6802'

Tutki tehtävää varten myös seuraavat scikit Learn -manuaalisivut:
* [Naive Bayes](https://scikit-learn.org/stable/modules/naive_bayes.html)
* [GaussianNB](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html#sklearn.naive_bayes.GaussianNB)



## Tehtävä 3.1
### Aihe: Wine Data Set (3 pistettä)

Toteuta tehtävä seuraavasti:

1. Tutustu *Wine Data Set* -aineistoon osoitteessa [https://archive.ics.uci.edu/ml/datasets/Wine](https://archive.ics.uci.edu/ml/datasets/Wine) .

2. Lataa aineisto ja aseta sarakkeiden nimet. 

3. Jaa aineisto koulutus- ja testausaineistoon käyttäen `scikit-learn`-kirjaston [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)-funktiota.  

4. Skaalaa koulutusaineiston sarakkeet välille 0...1 `scikit-learn`-kirjaston `preprocessing`-moduulin [minmax_scale](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.minmax_scale.html#sklearn.preprocessing.minmax_scale)-funktiolla.

* Tallenna koulutusaineisto ja sen luokkamuuttuja muuttujiin `X_train`, `y_train`,
* Tallenna testausaineisto vastaavasti muuttujiin `X_test`, `y_test`.

Aseta testausjoukon kooksi `33%` ja satunnaisuuden siemeneksi `2120`.

Vinkit:
* Vinkki: älä skaalaa luokkamuuttujaa.
* Vinkki: luokkamuuttujaa ei saa tallentaa `X`-muuttujiin.

Muuttujien nimeämisestä: 

* X-alkuiset muuttujat ovat ominaisuuksia, joita käytetään mallin syötteenä (input). 
* Y-alkuiset muuttujat ovat odotettuja tuloksia (labels).
* `X_train` ja `X_test` muuttujissa aineisto (dataset) on käytännössä vain jaettu (split) mallin koulutusta ja testaamista varten. 
* `y_train` ja `y_test` muuttujissa ovat vastaavasti odotetut tulokset (luokkamuuttujan).

Koulutus- ja testiaineiston ei tule sisältää samoja datarivejä. Eikä koskaan myöskään luokkamuuttujaa.

In [2]:
# TODO: Kirjoita toteutuksesi tähän soluun.
import pandas as pd
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

df = pd.read_csv('../../data/wine/wine.data', header=None)

# TODO: Sarakkeiden nimeäminen.
columns = [
    'cultivar', 'alcohol', 'malic-acid', 'ash', 'alcalinity-of-ash', 'magnesium', 'total-phenols',
    'flavanoids', 'nonflavanoid-phenols', 'proanthocyanins', 'color-intensity',
    'hue', 'od280-od315-diluted-wines', 'proline'
]

df.columns = columns

X = df.drop(columns=['cultivar'])
y = df['cultivar']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=2120)

mms = MinMaxScaler().fit(X_train)

X_train = pd.DataFrame(mms.transform(X_train), columns=X.columns)
X_test = pd.DataFrame(mms.transform(X_test), columns=X.columns)


In [13]:
# Tehtävän vastaukset. Huom! Älä muokkaa tätä solua, vaan aja se, kun olet suorittanut tehtävän. 
# Tallenna vastauksesi X_train- X_test-muuttujiin.
print(f'X_train columns: {X_train.columns}')
print(f'X_train: {X_train}')
print(f'X_test: {X_test}')

X_train columns: Index(['alcohol', 'malic-acid', 'ash', 'alcalinity-of-ash', 'magnesium',
       'total-phenols', 'flavanoids', 'nonflavanoid-phenols',
       'proanthocyanins', 'color-intensity', 'hue',
       'od280-od315-diluted-wines', 'proline'],
      dtype='object')
X_train:       alcohol  malic-acid       ash  alcalinity-of-ash  magnesium  \
0    0.444737    0.175153  0.491979           0.613402   0.152174   
1    0.139474    0.236253  1.000000           0.922680   0.532609   
2    0.342105    0.020367  0.315508           0.216495   0.717391   
3    0.344737    0.317719  0.588235           0.536082   0.304348   
4    0.721053    0.205703  0.705882           0.335052   0.489130   
..        ...         ...       ...                ...        ...   
114  0.394737    0.940937  0.684492           0.742268   0.282609   
115  0.507895    0.521385  0.529412           0.407216   0.391304   
116  0.581579    0.346232  0.807487           0.536082   0.521739   
117  0.431579    0.018330  

## Tehtävä 3.2
### Aihe: Gaussian Naive Bayes (2 pistettä)

Käytä `scikit-learn`-kirjastosta löytyvää *Gaussian Naive Bayes* -menetelmää aineiston luokittelemiseksi.

Ennnusta testiaineiston luokka. Tallenna ennusteet `y_pred`-muuttujaan.

In [4]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
# TODO: Kirjoita toteutuksesi tähän soluun.
gnb = GaussianNB()
y_pred = gnb.fit(X_train, y_train).predict(X_test)

In [5]:
# Tehtävän vastaukset. Huom! Älä muokkaa tätä solua, vaan aja se, kun olet suorittanut tehtävän. 
# Tallenna ennusteen tulokset y_pred-muuttujaan.
print(f'{len(y_pred)}')

59


## Tehtävä 3.3
### Aihe: Tarkkuuden mittaaminen (2 pistettä)

Hyödynnä `scikit-learn`-kirjaston `metrics`-moduulin funktioita ja laske luokittelutuloksesi **tarkkuus** sekä **sekaannusmatriisi**. Tallenna tarkkuus muuttujaan `acc` ja sekaannusmatriisi muuttujaan `cm`. 

* Vinkki: onnistumista voit mitata ainoastaan testiaineiston osalta!
* Vinkki: Tässä tehtävässä sekaannusmatriisi ja luokittelutuloksen tarkkuus luultavasti kohtalaisen hyviä (90 % -100 % välissä).

In [6]:
# TODO: Kirjoita toteutuksesi tähän soluun.
acc = cm = None
# TODO: luokittelutuloksesi tarkkuus sekä sekaannusmatriisi (confusion matrix)
acc = accuracy_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

In [7]:
# Tehtävän vastaukset. Huom! Älä muokkaa tätä solua, vaan aja se, kun olet suorittanut tehtävän. 
# Tallenna vastauksesi acc- ja cm-muuttujiin.
print(f'Accuracy: {acc}')
print(f'Confusion matrix:\n {cm}')

Accuracy: 1.0
Confusion matrix:
 [[24  0  0]
 [ 0 18  0]
 [ 0  0 17]]


## Tehtävä 3.4
### Aihe: Tulosten tulkinta (2 pistettä)

**Miksi tuloksia tulkitaan?** Tulosanalyysi on äärimmäisen tärkeää koneoppimisessa. Pelkkä (hyväkään) ratkaisu ei usein sellaisenaan riitä, vaan tulosten merkitys pitää ymmärtää "liiketoiminnan" kannalta.

Tulkitse sanallisesti saadut testitulokset.
Kuinka hyvin luokittelu onnistui? Pohdi myös miten valittu koneoppimismenetelmä toimi tässä tehtävässä.

* Vinkki: Jos sinulla riittää motivaatiota, niin kokeile ainakin jotakin muuta luokittelumenetelmää tässä tehtävässä. Tämä samalla varmistaa sen, että tässä tehtävässä saadut tulokset ovat "riittävän hyvät". 

In [8]:
from sklearn.neighbors import KNeighborsClassifier

k = 8
knn = KNeighborsClassifier(n_neighbors=k)
knn_pred = knn.fit(X_train, y_train).predict(X_test)
print(f'{len(knn_pred)}')

# TODO: Kirjoita toteutuksesi tähän soluun.
acc_knn = cm_knn = None
# TODO: luokittelutuloksesi tarkkuus sekä sekaannusmatriisi (confusion matrix)
acc_knn = accuracy_score(y_test, knn_pred)
cm_knn = confusion_matrix(y_test, knn_pred)

print(f'Accuracy: {acc_knn}')
print(f'Confusion matrix:\n {cm_knn}')

59
Accuracy: 0.9830508474576272
Confusion matrix:
 [[24  0  0]
 [ 0 18  0]
 [ 0  1 16]]


### Vastaus: Tulosten tulkinta

* TODO: TULOSTEN SANALLISTA TULKINTAA VASTATEN KYSYMYKSIIN, JOITA TEHTÄVÄSSÄ ESITETTY
* TODO: MUUTAKIN SAA TÄHÄN KOMMENTOIDA TEHTÄVÄÄN LIITTYEN.

Pohdinnan rivimäärä ei ole ratkaisevaa, mutta harvemmin riittää 1-2 riviä tekstiä, jos tehtävän tuloksia lähtee oikeasti miettimään! 

### Tämä lähti vähän käsistä. Muutenkin kiirettä moduulivaihdon takia, mutta alkoi kiinnostamaan. Toivottavasti edes vastasin oikeisiin kysymyksiin.

## Mitä teimme?

Ensin lisäsimme kolumneille otsikot. Tämän jälkeen tiputimme Xltä "oikean vastausrivin" ja lisäsimme sen y:lle
Muodostimme koulutus ja ennustus dataframet. Koulutukseen käytettiin 119 näytettä ja niitten perusteella
ennustettiin 59 näytettä. (66/33 split)

Käytimme X_train ja y_train dataa koulutukseen ja ennustimme X_test näytteille luokkamuuttujat. Tästä tuloksesta
tuli muuttuja y_pred ja tämä sai 59 ennustettua arvoa.

Käytimme accuracy_score ja confusion_matrix työkaluja arvioimaan mallin oikeellisuutta.
Työkalut vertasivat ennustettuja y_pred arvoja y_test arvoihin. y_test sisältää oikeat luokkamuuttujat.

Accurary = 1.0, eli kaikki ennustamamme näytteet olivat oikeissa luokkamuuttujissaan.
Confusion matrix osoitti vielä visuaalisesti meille tulosten oikeellisuuden.

Teimme samat temput knn koneoppimismallilla. (Toivottavasti tajusin vinkki tehtävän oikein)

knn antoi tarkkuudeksi 0.9830508474576272 ja confusion matrix varmisti tämän.

Confusion matrix:
 [[24  0  0]
 [ 0 18  0]
 [ 0  1 16]]

x-akseli = ennustettu luokka ja y-akseli = todellinen luokka.
Tämä tarkoittaa, että yksi luokan 3 arvo ennustettiin väärin luokkaan 2.

## Vastausten pohdinta

Tarkastetaan ensin yksinkertaisella esimerkillä miten gnb toimii:

Otetaan kaksi viiniä joistain alueista. Tässä on kaksi viiniä alueelta 2 ja kaksi viiniä alueelta 3.
Katson vain alkoholipitoisuuksia, sillä muuten tästä tulisi liian sekavaa ja pitkää.

Laskin kaavoja paperille mutta tähän kirjaan vastauksia.

2,12.37
2,12.04
3,12.86
3,12.88

Lasketaan keskiarvot ja keskihajonnat alueille:

keskiarvo alueella 2: 12.205
keskihajonta: 0.26

keskiarvo alueella 3: 12.87
keskihajonta: 0.01

P(12.5 | Alue2) = 0.808
P(12.5 | Alue3) = (1/0.02507)exp(-684.5)

Oma suomennos: P(12.5 | Alue3) = kuinka usein 12.5 alkoholipitoisuus tapahtuu kun Alue3 tapahtuu
= Kuinka todennäköistä on että 12.5 on alueella 3
(1/0.02507)exp(-684.5) = näin suuri negatiivinen exponentin arvo on *hyvin* lähellä nollaa,
joten voimme todeta että 12.5 alkoholipitoinen viini ei ole alueelta 3 ja tässä esimerkissä se olisi alueelta 2.

Jokaiselle viinin ominaisuudelle tehdään sama prosessi ja viinit osataan luokitella alueille kun kone on oppinut
alueiden ominaisuudet harjoitteludatalla.

## Miksi GNB toimii niin hyvin tässä?

- Datan eri ominaisuudet eivät ole vahvasti riippuvaisia toisistaan (ainakaan ymmärtääkseni, en ole viiniexpertti :D)

- Datan ominaisuudet ovat enimmäkseen keskittyneet keskiarvon ympärille (normaalijakauma)

- Selkeät erot luokkien välillä, tosin ei ehkä silmällä vilkaistuna

## Miksi GNB voitti KNN?

KNN ei ollut huono, se sai myös hyvin ennustettua testidatan luokat. En osaa kunnolla vastata tähän,
ehkä tulos ei ollut knn:n huonoutta, vaan yksinkertaisesti data sopi niin hyvin gnb:elle?
